kernel/init.c

```diff
@@ -550,6 +555,9 @@ FUNC_NORETURN void z_cstart(void)
 	/* do any necessary initialization of static devices */
 	z_device_state_init();
 
+#if CONFIG_PLATFORM_EARLY_INIT
+	platform_early_initialize();
+#endif
```

Comment:
wondering if this is early enough (considering EARLY exists and it is called just before entering the C domain). For example, z_device_state_init should be called after. Also, is `arch_kernel_init()` initializing something for the Kernel, or is it more like arch initial code (also required by early hooks)?

Reply:
right now I am just trying to make this work exactly the same way as before, architecture initialization might have other caveats with multi core and SMP and other memory related activities. We probably can eliminate EARLY and make this an arch hook as well. Anyways, the positioning is subject to optimizations and change, this is more about the concept...

Reply:
I think the concept is ok, but we should likely merge something that doesn't change in the near term, even if it implies some slight behavior changes.

Reply:
Why couldn't the invocation of a platform_early_initialize() just be a function run at PRE_KERNEL_1 priority 1 (or zero)?

Reply:
If you can hard-code the function call, you'll gain one init_entry less in ROM. (ok that's a very little amount)

SYS_INIT it meant to disappear, along with its hard-coded priority anyway.

The most important part of this PR is not where these functions are called: it could be easily fixed or rendered more flexible (whit what's going to replace SYS_INIT) since in kernel/init.c.  The API is what matters